backend/utils/email-service.js

```diff
@@ -0,0 +1,105 @@
+require('dotenv').config(); // For testing
+
+const axios = require('axios');
+const format = require('date-fns/format');
+const util = require('util');
+const supermarkets = require('../supermarkets');
+
+const formatSlotData = (slots, maxSlotsPerEmail) => {
+  const formattedSlots = [];
+
+  for (let i = 0; i < maxSlotsPerEmail; i++) {
+    const formattedSlot = {};
+
+    // If either of you can think of a nicer way to do this, let me know. I
+    // don't like the aesthetic.
+    formattedSlot.formattedDate = format(slots[i].start, 'EEEE, do LLLL');
+    formattedSlot.startTime = format(slots[i].start, 'k:mm');
+    formattedSlot.endTime = format(slots[i].end, 'k:mm');
+    if (slots[i].price) formattedSlot.price = slots[i].price; // TODO: Test this.
+
+    formattedSlots.push(formattedSlot);
+  }
+  return formattedSlots;
+}
+
+const buildSgPersonalization = (merchant, slots, address, url) => {
+  const maxSlotsPerEmail = slots.length > 5 ? 5 : slots.length;
+  const dynamicTemplateData = {
+    'btn-link': url,
+    more: slots.length > maxSlotsPerEmail ? true : false,
+    merchant: supermarkets[merchant],
+    slots: formatSlotData(slots, maxSlotsPerEmail)
+  };
+  const personalization = {
+    to: [{ email: address }],
+    dynamic_template_data: dynamicTemplateData
+  };
+  // console.log('personalization is:', personalization);
+  return personalization;
+};
+
+const buildSgPayload = (merchant, slots, addresses, url) => {
+  const sgPayload = {
+    from: { email: 'noreply@findadelivery.com' },
+    template_id:'d-ae627fe97d3c43209c1608fb43dfe7f0',
+    personalizations: addresses.map((address) => buildSgPersonalization(merchant, slots, address, url)) // SendGrid only lets you create 1000 of these per email
+  };
+  // console.log('sgPayload is:', sgPayload);
+  return sgPayload;
+}
+
+const defineAddresses = (addresses) => {
+  const primaryAddress = process.env.PERSONAL_EMAIL;
+
+  if (!primaryAddress && !addresses) throw { statusCode: 400, message: 'No recipient(s) found.' };
+  if (!primaryAddress) {
+    return addresses;
+  } else if (!addresses) {
+    return [primaryAddress];
+  } else {
+    return [primaryAddress, ...addresses];
+  }
+};
+
+const sendEmail = async (data) => {
+  // TODO: Build logger for debugging/testing
+  try {
+    if (!data) throw { statusCode: 400, message: 'No data passed.' };
+
+    console.log('Checking parameters...');
+    const requiredArgs = ['merchant', 'slots', 'url'];
```

Comment:
ignore me, seems like sometimes we'll be redirecting them to a specific page like /slots/2? is that why it's being passed dynamically?

Reply:
I assumed we would have a diff URL for each supermarket (and those URLs may change or be end up being dynamic) so instead of having to update a hard-coded value each time it changes, we could just pass it in each time.